### ISSUE TITLE
react-redux libdef causes errors with maybe-typed props

### ISSUE THREAD

**Issue by B-Reif** [flow-typed/flow-typed#2111]
I have some presentational components that expect a maybe-typed boolean `readOnly`. This is purposefully optional so consumers don't have to say `readOnly={false}`.

I have some container components that sometimes pass down a `readOnly` boolean, but theirs is not a maybe type. For example: 

https://flow.org/try/#0C4TwDgpgBACgTgezAZygXigbynCBDAEwHkA7AGxAH4AuKAIwQTPxKgF8AoUSKAZWDzAI8JKgzZchUhVoMmLdhw4BjBCWTAoAWxD9BEWnqEiU6LDnzFyIWgDM8ZZNE6r1mnSeS1PZiZek22roCQgB0klYU7EA

This pattern breaks in my codebase:
```
Error --------------------------------------------------- 

QuestionSelect.js:41:13
undefined [1] is incompatible with boolean [2].
   41| 	readOnly?: boolean
                   ^^^^^^^ [1]

References:
 QuestionSelectContainer.js:9:45
    9| type Props = { value: QuestionId, readOnly: boolean };
                                                   ^^^^^^^ [2]
```

Passing `readOnly` to a component expecting `readOnly?` should not error.

**Comment by lukeapage** [comment]
Your flow.org link doesn't error for me and it seems a incomplete example.

> I have some container components that sometimes pass down a readOnly boolean, but theirs is not a maybe type.

That doesn't seem right.. Can you give a full example?

**Comment by B-Reif** [comment]
You're right, I was in a bit of a hurry before. [Here's a slightly different but related example.](https://flow.org/try/#0PQKgUAkAKgFglgZwAQBcCeAHApkxSCGSCKArgEZIBmA9gE4FK1b4DGKSL1AthtQHZY+KAOTJOfSnADmJWvhRx+AOkgBJPo2ZsAtEwAmJAB6p4yADbVqAa3NwrOAKJmsXQSgDC-SVIA8nnvxuAHwqEAAUCFg4ZnACSABsAExI1JSarCi6WAaGAPoAbgCsSoYlSgBWCACUYCDAYOjYSP68AkKeEtJIALxIAN5IXGgAYiR8LABcSGFVPUGDcIbZSAC+ANxgYKCQsHiNOHiExORUdEgA7jDy6Tr6Rkh61FjI53AoMCY4eogY8iwf1DI5SwbFCu3MlhsRG4WHesSkSBi9ma3FabigmCwPgAJAAROCUSg+JwuNwdbx+bhBAA0SAAsriAApBIKQCJRRGxHAARgADIUUmkmBksjkCsVSqVKjU6g1MfT8BhsLQfAzmT1prNuvM8QSiS1Au0vNJaWrWWBxMRBmgAIJsRR8dzClBnXozObTAb7KYAIgAygB5OkOXI29xQVQBgByPtWVQ2cqadLQdMVyoDQJB7F6AyGo3GU3d2qQGeBbFWm2AwCQ6gIbBI+DMN0yd2M5ywwj0D2o0NccL4CKROAABuIBGwwnwSGYzLTcyMxiw42EDW0UFVhyottXsQ5DNg2A5aLQ6Bb+FahqmlVhaFMr8rVSm0zfS1n5m6tfMwn0wEhrfnJk1D1v1OahfTIfBaFjFYahgytwAgZMOFRQ12EeZ4+GEdgWEgnB8DIagSHYd4cCYUhaA0fYkBgG8sFpBAezeJASEiZBhy4RZsmHXANDeZAMBPDAiGkPh5FkLAVD-P9YDw+tGzrBR+A4Z0zkoE8uGbUV7jI2Q+GQfANFfUFfyk1QkCwfcsyoCxziQPhqGInsbxPegaKYNYGA4pYuyohAYEIswuzIHBJDiGh6AM5ip0iLsdIopB8kbEgJJMv9hjOCz8B4ZxaThZB0OQez2GcugJlqeoqOTUsNQGGgwKA4svOWFYzz09ghlLO80GqnNQMLT9PRg1YkCrJAAHU6BsABCMAgA)



This example is related to the dispatch props but exhibits a similar behavior: a component expects a certain type, a container supplies something of a valid subtype, and flow incorrectly errors out.

In my first example, the container's `readOnly: boolean` should work as a subtype of the component's `readOnly?: boolean`. In this example, the container's `myFunc: () => Object` should work as a subtype of the component's `myFunc: () => mixed`.

Something about the `$Diff<ElementConfig>` type is causing a problem where only an exact type match is accepted, and valid subtypes are erroring.

**Comment by AndrewSouthpaw** [comment]
Makes sense, although I don't think Flow is incorrect per se, because a boolean is different from a void in the eyes of Flow. Could you simply do something like in MSTP pass along `{ readOnly: !!ownProps.readOnly }`?

**Comment by B-Reif** [comment]
The Container is injecting `readOnly` in that case, so it's just coming from a selector.

The point is that the component expects `boolean | void` and then errors on `boolean`.

**Comment by AndrewSouthpaw** [comment]
Mm... I don't think so. See [this try link](https://flow.org/try/#0PTAEBMHsFMGdQHaQC6gO6QE4GsBQyBPAB2lAAVNIj4BeUAb1E2gENwB5BAGwIH4AuUACNIkLqwSgAvvmKkAyshbJoFKrQZNWHbgUEixE6blwBjSAlioAtgTXVB9jY2ZtOPQQDMWXWKRnmljYEisrQgqEqTqB0LtrueqC2TgB0rjo8xkA), as I think it illustrates what you're trying to do as I understand it.

Could be totally misunderstanding the problem though so I'm all ears and curious to debug.

**Comment by B-Reif** [comment]
This try link specifies the types correctly but assigns the actual value backward. 
```js
const myProps: Props = { readOnly: false }
const myState: StateProps = { readOnly: myProps.readOnly }
```
should be 
```js
const myState: StateProps = { readOnly: false }
const myProps: Props = { readOnly: myState.readOnly }
```
The state informs the component, not the other way around. Make this change in the REPL and the example works with no errors.

With react-redux, this causes an error. I pushed a minimal reproduction at https://github.com/B-Reif/flow-typed-react-redux-repro.

Running flow on this code produces this error:
```
Error ┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈┈ src/InputContainer.js:6:9

Cannot return object literal because undefined [1] is incompatible with boolean [2].

     src/InputContainer.js
     3│ import Input from "./Input";
     4│
     5│ const mapStateToProps = (_, state) => {
 [2] 6│         return { readOnly: true };
     7│ };
     8│
     9│ export default connect(mapStateToProps)(Input);

     src/Input.js
 [1] 4│ type Props = { readOnly?: boolean };
```

Let me know if there's anything else I can provide.

**Comment by AndrewSouthpaw** [comment]
@villesau Care to weigh in?

**Comment by villesau** [comment]
@B-Reif Would you mind providing example repo? Something which uses the real libdef and component to illustrate the problem? I guess that would help to understand the case a bit better.

**Comment by B-Reif** [comment]
Yep, already made it, see: https://github.com/B-Reif/flow-typed-react-redux-repro

**Comment by villesau** [comment]
Ok now I got it, thanks! _Looks_ like the flow check is somewhat other way around in this case. It might be possible to solve with a bit of reordering of the types in the libdefs. I could check it out once I have time, but can't promise anything.

**Comment by thenanyu** [comment]
I think this is the same issue as https://github.com/flowtype/flow-typed/issues/1997
We're running into an issue where doing `$Diff` causes the flow error to look like it's being evaluated backwards.

https://flow.org/try/#0C4TwDgpgBACgTgezAZygXigbygMwQgLigDsBXAWwCMI4AaKSgQwC8B+I5YOAS2IHMoAXwDcAKFCQoASWIArCAGNgEACbwkqDNibMOXXgJHjw0AEoQAjqW5xV6lOigASACLccOADz3k9GfKU7RBQAPjFRBQRiTihlAA9gInMrGyCNR0w8QigARkEgA

This may be related to the disclaimer here: https://flow.org/en/docs/types/utilities/#toc-diff
> Note that `$Diff<A, B>` will error if the object you are removing properties from does not have the property being removed, i.e. if B has a key that doesn’t exist in A
